File: Compilers - Principles, Techniques, and Tools/Chapter 3: Lexical Analysis/3.4 Transition Diagrams/Exercise 3.4.9/main.cc
#include <string_view>
#include <vector>

#include <fmt/core.h>

std::string fibonacci_string(size_t n) {
  static std::vector<std::string> mem{"b", "a"};
  if (n > mem.size()) mem.reserve(n);
  for (size_t k = mem.size(); k < n; ++k)
    mem.emplace_back(mem[k - 1] + mem[k - 2]);
  return mem[n - 1];
}

std::vector<size_t> failure_function(std::string_view p) {
  std::vector<size_t> failures(p.size());
  for (size_t s = 0, k = 1; k != p.size(); ++k) {
    while (s != 0 && p[s] != p[k]) s = failures[s - 1];
    if (p[s] == p[k]) s = s + 1;
    failures[k] = s;
  }
  return failures;
}

size_t integer_length(uintmax_t n) {
  return std::to_string(n).size();
}

void println_markdown_failure_function_table(const std::string_view sv) {
  const size_t length = sv.size();
  const auto failures = failure_function(sv);
  fmt::print("\n|  *s* |");
  for (size_t k = 0; k != length; ++k)
    fmt::print("{:^{}}|", sv[k], integer_length(failures[k]) + 2);
  fmt::print("\n|:----:|");
  for (size_t k = 0; k != length; ++k)
    fmt::print(":{:-^{}}:|", "", integer_length(failures[k]));
  fmt::print("\n|*f(s)*|");
  for (size_t k = 0; k != length; ++k) fmt::print(" {} |", failures[k]);
  fmt::print("\n");
}

int main() {
  println_markdown_failure_function_table(fibonacci_string(6));
  println_markdown_failure_function_table(fibonacci_string(7));
}

File: Compilers - Principles, Techniques, and Tools/Chapter 3: Lexical Analysis/3.4 Transition Diagrams/Exercise 3.4.9/main.rs
fn fibonacci_string(n: usize) -> String {
    let mut mem: once_cell::sync::Lazy<Vec<String>> =
        once_cell::sync::Lazy::new(|| vec!["b".into(), "a".into()]);
    if n > mem.len() {
        mem.reserve(n);
    }
    for k in mem.len()..n {
        let len = mem[k - 1].len() + mem[k - 2].len();
        let mut next = String::with_capacity(len);
        next.push_str(&mem[k - 1]);
        next.push_str(&mem[k - 2]);
        mem.push(next);
    }
    return mem[n - 1].clone();
}

fn failure_function(p: impl AsRef<str>) -> Vec<usize> {
    let p = p.as_ref().chars().collect::<Vec<_>>();
    let mut failure = vec![0usize; p.len()];
    let mut s = 0;
    for k in 1..p.len() {
        while s != 0 && p[s] != p[k] {
            s = failure[s - 1];
        }
        if p[s] == p[k] {
            s = s + 1;
        }
        failure[k] = s;
    }
    return failure;
}

fn digits_length(n: usize) -> usize {
    n.to_string().len()
}

fn print_markdown_failure_function_table(s: impl AsRef<str>) {
    let failure = failure_function(s.as_ref());
    print!("\n|  *s* |");
    for (k, c) in s.as_ref().char_indices() {
        print!("{:^1$}|", c, digits_length(failure[k]) + 2);
    }
    print!("\n|:----:|");
    for &k in failure.iter() {
        print!(":{:-^1$}:|", "", digits_length(k));
    }
    print!("\n|*f(s)*|");
    for k in failure.iter() {
        print!(" {} |", k);
    }
    println!();
}

fn main() {
    print_markdown_failure_function_table(fibonacci_string(6));
    print_markdown_failure_function_table(fibonacci_string(7));
}
